### PR TITLE
Update Onboard Resource

### PR DESCRIPTION
**Pull request by cyy0714**
1. add endpoints for delete existing anomalies and regenerate anomalies

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java [hunk 1]
```diff
@@ -217,4 +222,100 @@ public Boolean cloneAnomalyInstances(Long srcId, Long destId) {
...
+   */
+  @POST
+  @Path("function/{id}/regenerateAnomaliesInRange")
+  public String regenerateAnomaliesInRange(@PathParam("id") String id,
+      @QueryParam("detectionHost") String detectionHost,
+      @QueryParam("detectionHost") int detectionPort,
+      @QueryParam("monitoringWindowStartTime") long monitoringWindowStartTime,
+      @QueryParam("monitoringWindowEndTime") long monitoringWindowEndTime,
+      @QueryParam("isForceBackfill") @DefaultValue("false") String isForceBackfill) throws Exception {
+    long functionId = Long.valueOf(id);
```
Comment (puneetjaiswal) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: should not this take care of deleting the existing anomalies for this time window first before re - generating ?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java [hunk 2]
```diff
@@ -217,4 +222,100 @@ public Boolean cloneAnomalyInstances(Long srcId, Long destId) {
...
+
+  /**
+   * Breaks down the given range into consecutive monitoring windows as per function definition
+   * Regenerates anomalies for each window separately
+   * @param id anomaly function id
+   * @param detectionHost the host of anomaly detection function
+   * @param detectionPort the port of anomaly detection function
+   * @param monitoringWindowStartTime The start time of the monitoring window (in milli-second)
+   * @param monitoringWindowEndTime The start time of the monitoring window (in milli-second)
+   * @param isForceBackfill determine if it should backfill
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: It should be: false to resume backfill from last left off

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java [hunk 3]
```diff
@@ -217,4 +222,100 @@ public Boolean cloneAnomalyInstances(Long srcId, Long destId) {
...
+   * @return
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: HTTP response of this request

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java [hunk 4]
```diff
@@ -217,4 +222,100 @@ public Boolean cloneAnomalyInstances(Long srcId, Long destId) {
...
+      mergedAnomaliesDeleted = deleteMergedResults(mergedResults);
+    }
+
+    LOG.info("Deleted {} merged anomalies", mergedAnomaliesDeleted);
+    return mergedAnomaliesDeleted;
+  }
+
+  // Delete merged anomalies from mergedAnomalyResultDAO
+  private int deleteMergedResults(List<MergedAnomalyResultDTO> mergedResults) {
+    LOG.info("Deleting merged results");
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: Could be more specific: merged anomaly results

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java [hunk 5]
```diff
@@ -217,4 +222,100 @@ public Boolean cloneAnomalyInstances(Long srcId, Long destId) {
...
+   * @param monitoringWindowEndTime The start time of the monitoring window (in milli-second)
+   * @param isForceBackfill determine if it should backfill
+   * @return
+   * @throws Exception
+   */
+  @POST
+  @Path("function/{id}/regenerateAnomaliesInRange")
+  public String regenerateAnomaliesInRange(@PathParam("id") String id,
+      @QueryParam("detectionHost") String detectionHost,
+      @QueryParam("detectionHost") int detectionPort,
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: I would suggest that you test this endpoint against 'dashobard' and 'anomalydetector' on your local host before you merge this change.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java [hunk 6]
```diff
@@ -217,4 +222,100 @@ public Boolean cloneAnomalyInstances(Long srcId, Long destId) {
...
+  }
+
+  // Delete merged anomalies from mergedAnomalyResultDAO
+  private int deleteMergedResults(List<MergedAnomalyResultDTO> mergedResults) {
+    LOG.info("Deleting merged results");
+    int mergedAnomaliesDeleted = 0;
+    for (MergedAnomalyResultDTO mergedResult : mergedResults) {
+      // Delete raw anomalies of the merged anomaly
+      List<RawAnomalyResultDTO> rawAnomalyResultDTOs = mergedResult.getAnomalyResults();
+      //deleteRawResults(rawAnomalyResultDTOs);
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: Should remove raw anomaly results before before regenerate. Another place need to modify is the clone method, when clone merged anomalies, we should either reset their 'raw anomaly ids to null', or clone the raw anomaly result together with merged anomalies, otherwise their might be still a problem. @cyenjung is this a valid concern?

Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: Good catch. The raw anomalies that belong to the "to be deleted" merged anomalies should also be removed. Otherwise, anomaly function might not generate new raw anomalies whose range is located inside the range of any existing raw anomalies. Please delete the raw anomalies as well.

Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: In the cloneAnomalyInstances, 
>for (MergedAnomalyResultDTO mergedAnomalyResultDTO : mergedAnomalyResultDTOs) {
>      long oldId = mergedAnomalyResultDTO.getId();
>      mergedAnomalyResultDTO.setId(null);  // clean the Id, then will create a new Id when save
>      mergedAnomalyResultDTO.setRawAnomalyIdList(null);
>      mergedAnomalyResultDTO.setFunctionId(destId);
>      long newId = mergedAnomalyResultDAO.save(mergedAnomalyResultDTO);
>      LOG.debug("clone merged anomaly {} to {}", oldId, newId);
>    }


It doesn't copy the raw anomaly to the new merged results.

Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: Issue Fixed

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 7]
```diff
@@ -190,4 +201,92 @@ public Response computeSeverity(@NotNull @QueryParam("collection") String collec
...
+   * @param monitoringWindowStartTime The start time of the monitoring window (in milli-second)
+   * @param monitoringWindowEndTime The start time of the monitoring window (in milli-second)
+   * @param isForceBackfill false to resume backfill from the latest left off
+   * @return HTTP response of this request
+   * @throws Exception
+   */
+  //private final String detectionHost = "localhost";
+  //private final int detectionPort = 1867;
+  @POST
+  @Path("/{id}/regenerateAnomaliesInRange")
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: There's a method few lines above in this file, called backfill. It is doing exactly what this method is doing. Please make sure we don't duplicate this functionality. Either remove the backfill method, or make this regenerate method call backfill method.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java [hunk 8]
```diff
@@ -201,20 +214,165 @@ public Boolean cloneAnomalyInstances(Long srcId, Long destId) {
...
+   * 2. If a raw anomaly belongs to a merged anomaly whose start time is located in the given
+   * time ranges, then it is deleted regardless its start time.
+   *
+   * If monitoringWindowStartTime is not given, then start time is set to 0.
+   * If monitoringWindowEndTime is not given, then end time is set to Long.MAX_VALUE.
+   * @param monitoringWindowStartTime The start time of the monitoring window (in milli-second)
+   * @param monitoringWindowEndTime The start time of the monitoring window (in milli-second)
+   */
+  @POST
+  @Path("function/{id}/deleteExistingAnomalies")
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/OnboardResource.java]: This method looks exactly like the cleanup method in the CleanupAndRegenerateAnomaliesTool. Please have a look at CleanupAndRegenerateAnomaliesTool, and remove the duplicate code frmo there. Make the tool call these methods. This file can be the source of the functionality, and the Tool can be another way of accessing them.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 9]
```diff
@@ -190,4 +165,88 @@ public Response computeSeverity(@NotNull @QueryParam("collection") String collec
...
   }
+
+  /**
+   * Breaks down the given range into consecutive monitoring windows as per function definition
+   * Regenerates anomalies for each window separately
+   *
+   * As the anomaly result regeneration is a heavy job, we move the function from Dashboard to worker
+   * @param id anomaly function id
+   * @param startTimeIso The start time of the monitoring window (in ISO Format), ex: 2016-5-23T00:00:00Z
+   * @param endTimeIso The start time of the monitoring window (in ISO Format)
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: Typo: The end time.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 10]
```diff
@@ -190,4 +165,88 @@ public Response computeSeverity(@NotNull @QueryParam("collection") String collec
...
+  @Path("/{id}/generateAnomaliesInRange")
+  public Response generateAnomaliesInRange(@PathParam("id") String id,
+      @QueryParam("start") String startTimeIso,
+      @QueryParam("end") String endTimeIso,
+      @QueryParam("force") @DefaultValue("false") String isForceBackfill) throws Exception {
+    long functionId = Long.valueOf(id);
+    AnomalyFunctionDTO anomalyFunction = anomalyFunctionDAO.findById(functionId);
+    if (anomalyFunction == null) {
+      return Response.noContent().build();
+
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: Remove this empty line.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 11]
```diff
@@ -190,4 +165,88 @@ public Response computeSeverity(@NotNull @QueryParam("collection") String collec
...
+    }
+
+    startTime = ISODateTimeFormat.dateTimeParser().parseDateTime(startTimeIso);
+    endTime = ISODateTimeFormat.dateTimeParser().parseDateTime(endTimeIso);;
+
+    if(startTime.isAfter(endTime)){
+      throw new IllegalArgumentException(String.format(
+          "[functionId %s] Monitoring start time is after monitoring end time", id));
+    }
+    if(endTime.isAfterNow()){
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: This is a very good checking! Could you improve it further and check the max date time of the dataset instead of the current time.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 12]
```diff
@@ -190,4 +165,88 @@ public Response computeSeverity(@NotNull @QueryParam("collection") String collec
...
+    //  Check if the raw anomaly results have been cleaned up before regeneration
+    List<RawAnomalyResultDTO> rawResults =
+        rawAnomalyResultDAO.findAllByTimeAndFunctionId(startTime.getMillis(), endTime.getMillis(), functionId);
+    if(CollectionUtils.isNotEmpty(rawResults)){
+      throw new IllegalArgumentException(String.format(
+          "[functionId {}] Raw anomaly results should be cleaned up before regeneration", id));
+    }
+
+    // Check if the anomaly function is active
+    if (!anomalyFunction.getIsActive()) {
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: Check this one before all other checkings?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 13]
```diff
@@ -190,4 +165,88 @@ public Response computeSeverity(@NotNull @QueryParam("collection") String collec
...
+      throw new IllegalArgumentException(String.format("Skipping deactivated function %s", id));
+    }
+
+    String response = null;
+
+    DateTime innerStartTime = startTime;
+    DateTime innerEndTime = endTime;
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: What's the purpose of coping these two redundant variables?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/views/function/SplineRegressionAnomalyTimeSeriesView.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
+package com.linkedin.thirdeye.anomaly.views.function;
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/views/function/SplineRegressionAnomalyTimeSeriesView.java]: Is this class related to onboarding? Could you separate it to another branch?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/views/function/SplineRegressionAnomalyTimeSeriesView.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    if(knownAnomalies != null && !knownAnomalies.isEmpty()) {
+      for (RawAnomalyResultDTO anomaly : knownAnomalies) {
+        AnomalyFeedbackDTO feedbackDTO = anomaly.getFeedback(); // null if no user feedback
+        TimeRange anomalyTimeRange = new TimeRange(anomaly.getStartTime(), anomaly.getEndTime());
+        AnomalyResult result = new AnomalyResult((feedbackDTO == null), anomalyTimeRange, anomaly.getScore(), Math.abs(anomaly.getWeight()), AnomalyResult.AnomalyMethod.SPLINEREGRESSION, properties);
+        anomalyHistory.add(result);
+      }
+    }
+
+    MapMetricTimeSeries series = new MapMetricTimeSeries();
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/views/function/SplineRegressionAnomalyTimeSeriesView.java]: I am not sure if it is a good idea to directly access anomaly detection library here. We need a property interface to communicate this internal libraries.

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/tools/CleanupAndRegenerateAnomaliesTool.java [hunk 16]
```diff
@@ -1,287 +0,0 @@
-package com.linkedin.thirdeye.tools;
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/tools/CleanupAndRegenerateAnomaliesTool.java]: Don't delete this utility class. We still need this tool to do adhoc cleaning and regeneration of anomalies. Since you have copied the code to OnboardingResource, you could invoke those methods from this class. Otherwise, we have to start an anomaly application in order to clean or regenerate anomalies, which is really inconvenient.

File: thirdeye/thirdeye-pinot/src/main/resources/assets/js/lib/self-service.js [hunk 17]
```diff
@@ -364,6 +364,7 @@ function addSelfServiceListeners() {
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/resources/assets/js/lib/self-service.js]: Move to another branch. Onboarding and enabling Spline Regression on TE should be two separated tasks.

File: thirdeye/thirdeye-pinot/src/main/resources/com/linkedin/thirdeye/dashboard/views/tabs/self-service-tab/anomaly-function-form.ftl [hunk 18]
```diff
@@ -169,7 +169,7 @@
         <div class="uk-form-row uk-margin-top">
 
             <!-- ** WEEK_OVER_WEEK_RULE & MIN_MAX_THRESHOLD PROPERTIES ** -->
-            <div class="WEEK_OVER_WEEK_RULE-fields MIN_MAX_THRESHOLD-fields function-type-fields">
+            <div class="dimension-selection-fields function-type-fields">
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/resources/com/linkedin/thirdeye/dashboard/views/tabs/self-service-tab/anomaly-function-form.ftl]: Move this change to another branch.

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/tools/CleanupAndRegenerateAnomaliesConfig.java [hunk 19]
```diff
@@ -1,72 +0,0 @@
-package com.linkedin.thirdeye.tools;
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/tools/CleanupAndRegenerateAnomaliesConfig.java]: Please do not remove this utility class.

File: thirdeye/thirdeye-pinot/src/main/resources/assets/js/lib/self-service.js (hunk 17 above)
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/resources/assets/js/lib/self-service.js]: Agreed, let's make smaller batch of changes and merge more frequently.
